File: src/main/java/org/academy/api/common/profiler/ProfileDump.java
package org.academy.api.common.profiler;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;

public final class ProfileDump {
    private static final DateTimeFormatter TIME_FORMAT = DateTimeFormatter.ofPattern("yyyyMMdd-HHmmss");

    private ProfileDump() {
    }

    public static String timestamp() {
        return LocalDateTime.now().format(TIME_FORMAT);
    }

    public static String dumpZones(ZoneSnapshot snapshot, int maxDepth) {
        var sb = new StringBuilder();
        sb.append("== AcademyCraft Zone Profile [thread=")
                .append(snapshot.getThreadName())
                .append("] ==\n");
        var rootTotal = snapshot.getRootTotalNs();
        sb.append("Total: ")
                .append(String.format("%.2f", rootTotal / 1e6))
                .append(" ms\n");
        appendZoneNode(sb, snapshot, ZoneProfiler.ROOT, 0, maxDepth);
        return sb.toString();
    }

    public static String dumpZones(ZoneSnapshot snapshot) {
        return dumpZones(snapshot, 8);
    }

    private static void appendZoneNode(StringBuilder sb, ZoneSnapshot snapshot, String path, int depth, int maxDepth) {
        if (depth > maxDepth) {
            return;
        }
        for (var child : snapshot.childrenOf(path)) {
            var pct = snapshot.parentPercent(child);
            var rootTotal = snapshot.getRootTotalNs();
            var gpct = rootTotal > 0 ? child.totalNs() * 100.0 / rootTotal : 0.0;
            sb.append("  ".repeat(depth))
                    .append(child.name())
                    .append(" - ")
                    .append(String.format("%.2f%% / %.2f%%", pct, gpct))
                    .append(" - ")
                    .append(String.format("%.2f ms", child.getTotalMs()))
                    .append(" (self ")
                    .append(String.format("%.2f ms", child.getSelfMs()))
                    .append(") - ")
                    .append(child.count())
                    .append(" calls - max ")
                    .append(String.format("%.2f ms", child.getMaxMs()))
                    .append('\n');
            if (depth < maxDepth) {
                appendZoneNode(sb, snapshot, child.path(), depth + 1, maxDepth);
            }
        }
    }

    public static String dumpSampler(SamplerSnapshot snapshot, int limit) {
        var sb = new StringBuilder();
        sb.append("== AcademyCraft Sampling Profile ==\n")
                .append("Total samples: ")
                .append(snapshot.totalSamples())
                .append("  Duration: ")
                .append(String.format("%.1f", snapshot.durationSeconds()))
                .append(" s\n");
        for (var view : snapshot.threads().values()) {
            sb.append("-- Thread: ").append(view.name())
                    .append(" (").append(view.samples()).append(" samples)\n");
            Map<String, Long> self = new HashMap<>();
            collectSelf(view.root(), self);
            List<Map.Entry<String, Long>> top = new ArrayList<>(self.entrySet());
            top.sort((a, b) -> Long.compare(b.getValue(), a.getValue()));
            var total = view.root().samples();
            for (var i = 0; i < Math.min(limit, top.size()); i++) {
                var entry = top.get(i);
                var pct = total > 0 ? entry.getValue() * 100.0 / total : 0.0;
                sb.append("  ")
                        .append(entry.getKey())
                        .append(" - ")
                        .append(String.format("%.2f%%", pct))
                        .append(" (")
                        .append(entry.getValue())
                        .append(" self samples)\n");
            }
        }
        return sb.toString();
    }

    public static String dumpSampler(SamplerSnapshot snapshot) {
        return dumpSampler(snapshot, 30);
    }

    private static void collectSelf(SampledNode node, Map<String, Long> map) {
        if (!"<root>".equals(node.label())) {
            map.merge(node.label(), node.selfSamples(), Long::sum);
        }
        for (var child : node.children()) {
            collectSelf(child, map);
        }
    }

    public static String status(ProfilerSnapshot snapshot) {
        var sb = new StringBuilder();
        sb.append("== AcademyCraft Profiler Status ==\n")
                .append("Zone capture: ").append(snapshot.isZonesEnabled() ? "ON" : "OFF").append('\n');
        var sampler = snapshot.getSampler();
        sb.append("Sampling: ").append(snapshot.isSampling() ? "ON" : "OFF");
        if (snapshot.isSampling()) {
            sb.append(" (paused: ").append(snapshot.isSamplingPaused() ? "yes" : "no").append(')');
            if (sampler != null) {
                sb.append("\n  samples: ").append(sampler.totalSamples())
                        .append("  duration: ").append(String.format("%.1f", sampler.durationSeconds()))
                        .append(" s");
            }
        }
        sb.append("\nThreads:\n");
        for (var ref : AcademyProfiler.samplerThreads()) {
            sb.append("  ").append(ref.getName()).append(" (id ").append(ref.getId()).append(") ")
                    .append(ref.isEnabled() ? "enabled" : "disabled").append('\n');
        }
        return sb.toString();
    }

    public static String zonesText(ProfilerSnapshot snapshot, String threadName, int maxDepth) {
        var zones = snapshot.getZones();
        if (zones.isEmpty()) {
            var state = snapshot.isZonesEnabled() ? "capture ON" : "capture OFF";
            return "== Zone Profile ==\nNo zone data (" + state + "). " +
                    "Start capture, then trigger the effect, then stop.\n";
        }
        Map<String, ZoneSnapshot> targets;
        if (threadName != null) {
            targets = new LinkedHashMap<>();
            for (var entry : zones.entrySet()) {
                if (threadName.equals(entry.getKey())) {
                    targets.put(entry.getKey(), entry.getValue());
                }
            }
            if (targets.isEmpty()) {
                return "== Zone Profile ==\nThread '" + threadName + "' not found. Available: " +
                        String.join(", ", zones.keySet()) + ".\n";
            }
        } else {
            targets = zones;
        }
        var sb = new StringBuilder();
        for (var entry : targets.entrySet()) {
            sb.append(dumpZones(entry.getValue(), maxDepth)).append('\n');
        }
        return sb.toString();
    }

    public static String samplerText(ProfilerSnapshot snapshot, int topN) {
        var sampler = snapshot.getSampler();
        if (sampler == null) {
            var state = snapshot.isSampling() ? "sampling ON" : "sampling OFF";
            return "== Sampling Profile ==\nNo sampler data (" + state + "). " +
                    "Start sampling, then trigger the effect, then stop.\n";
        }
        return dumpSampler(sampler, topN);
    }
}

File: src/main/kotlin/org/academy/api/common/profiler/ProfileDump.kt
package org.academy.api.common.profiler

import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

/**
 * 将剖析数据导出为文本报告（命令 / 文件导出用）。
 */
object ProfileDump {
    private val TIME_FORMAT = DateTimeFormatter.ofPattern("yyyyMMdd-HHmmss")

    @JvmStatic
    fun timestamp(): String = LocalDateTime.now().format(TIME_FORMAT)

    @JvmStatic
    fun dumpZones(snapshot: ZoneSnapshot, maxDepth: Int = 8): String {
        val sb = StringBuilder()
        sb.append("== AcademyCraft Zone Profile [thread=")
            .append(snapshot.threadName)
            .append("] ==\n")
        val rootTotal = snapshot.rootTotalNs
        sb.append("Total: ")
            .append("%.2f".format(rootTotal / 1e6))
            .append(" ms\n")
        appendZoneNode(sb, snapshot, ZoneProfiler.ROOT, 0, maxDepth)
        return sb.toString()
    }

    private fun appendZoneNode(
        sb: StringBuilder,
        snapshot: ZoneSnapshot,
        path: String,
        depth: Int,
        maxDepth: Int,
    ) {
        if (depth > maxDepth) return
        for (child in snapshot.childrenOf(path)) {
            val pct = snapshot.parentPercent(child)
            val rootTotal = snapshot.rootTotalNs
            val gpct = if (rootTotal > 0) child.totalNs * 100.0 / rootTotal else 0.0
            sb.append("  ".repeat(depth))
                .append(child.name)
                .append(" - ")
                .append("%.2f%% / %.2f%%".format(pct, gpct))
                .append(" - ")
                .append("%.2f ms".format(child.totalMs))
                .append(" (self ")
                .append("%.2f ms".format(child.selfMs))
                .append(") - ")
                .append(child.count)
                .append(" calls - max ")
                .append("%.2f ms".format(child.maxMs))
                .append('\n')
            if (depth < maxDepth) {
                appendZoneNode(sb, snapshot, child.path, depth + 1, maxDepth)
            }
        }
    }

    @JvmStatic
    fun dumpSampler(snapshot: SamplerSnapshot, limit: Int = 30): String {
        val sb = StringBuilder()
        sb.append("== AcademyCraft Sampling Profile ==\n")
            .append("Total samples: ")
            .append(snapshot.totalSamples)
            .append("  Duration: ")
            .append("%.1f".format(snapshot.durationSeconds))
            .append(" s\n")
        for (view in snapshot.threads.values) {
            sb.append("-- Thread: ").append(view.name).append(" (").append(view.samples).append(" samples)\n")
            val self = HashMap<String, Long>()
            collectSelf(view.root, self)
            val top = self.entries.sortedByDescending { it.value }.take(limit)
            for ((label, count) in top) {
                val total = view.root.samples
                val pct = if (total > 0) count * 100.0 / total else 0.0
                sb.append("  ")
                    .append(label)
                    .append(" - ")
                    .append("%.2f%%".format(pct))
                    .append(" (")
                    .append(count)
                    .append(" self samples)\n")
            }
        }
        return sb.toString()
    }

    private fun collectSelf(node: SampledNode, map: MutableMap<String, Long>) {
        if (node.label != "<root>") {
            map.merge(node.label, node.selfSamples, Long::plus)
        }
        for (child in node.children) {
            collectSelf(child, map)
        }
    }

    /** 命令用：当前采集状态总览。 */
    @JvmStatic
    fun status(snapshot: ProfilerSnapshot): String {
        val sb = StringBuilder()
        sb.append("== AcademyCraft Profiler Status ==\n")
            .append("Zone capture: ").append(if (snapshot.zonesEnabled) "ON" else "OFF").append('\n')
        val sampler = snapshot.sampler
        sb.append("Sampling: ").append(if (snapshot.sampling) "ON" else "OFF")
        if (snapshot.sampling) {
            sb.append(" (paused: ").append(if (snapshot.samplingPaused) "yes" else "no").append(')')
            if (sampler != null) {
                sb.append("\n  samples: ").append(sampler.totalSamples)
                    .append("  duration: ").append("%.1f".format(sampler.durationSeconds)).append(" s")
            }
        }
        sb.append("\nThreads:\n")
        for (ref in AcademyProfiler.samplerThreads()) {
            sb.append("  ").append(ref.name).append(" (id ").append(ref.id).append(") ")
                .append(if (ref.enabled) "enabled" else "disabled").append('\n')
        }
        return sb.toString()
    }

    /** 命令用：zone 树详细查看（可按线程过滤）。 */
    @JvmStatic
    fun zonesText(snapshot: ProfilerSnapshot, threadName: String?, maxDepth: Int): String {
        val zones = snapshot.zones
        if (zones.isEmpty()) {
            val state = if (snapshot.zonesEnabled) "capture ON" else "capture OFF"
            return "== Zone Profile ==\nNo zone data ($state). " +
                    "Start capture, then trigger the effect, then stop.\n"
        }
        val targets = if (threadName != null) zones.filterKeys { it == threadName } else zones
        if (targets.isEmpty()) {
            return "== Zone Profile ==\nThread '$threadName' not found. Available: " +
                    zones.keys.joinToString(", ") + ".\n"
        }
        val sb = StringBuilder()
        for ((name, zoneSnapshot) in targets) {
            sb.append(dumpZones(zoneSnapshot, maxDepth)).append('\n')
        }
        return sb.toString()
    }

    /** 命令用：采样式方法级 top 查看。 */
    @JvmStatic
    fun samplerText(snapshot: ProfilerSnapshot, topN: Int): String {
        val sampler = snapshot.sampler
        if (sampler == null) {
            val state = if (snapshot.sampling) "sampling ON" else "sampling OFF"
            return "== Sampling Profile ==\nNo sampler data ($state). " +
                    "Start sampling, then trigger the effect, then stop.\n"
        }
        return dumpSampler(sampler, topN)
    }
}
